servicetalk-http-netty/src/main/java/io/servicetalk/http/netty/NettyHttpServer.java

```diff
@@ -273,41 +271,39 @@ public String toString() {
-                                    // determined to be 0. In that case we should conservatively use End and rely upon
-                                    // adjustForMissingBoundaries to accommodate if more data comes. Otherwise the
-                                    // FlushStrategy may not trigger a flush (e.g. flushOnEnd) and if so the response
-                                    // won't actually be written.
-                                    return contentLength > 0 ||
-                                            (contentLength < 0 && isTransferEncodingChunked(metadata.headers())) ?
-                                            Start : End;
-                                }
-                                if (itemWritten instanceof Buffer) {
-                                    return contentLength > 0 &&
-                                            (contentLength -= ((Buffer) itemWritten).readableBytes()) <= 0 ?
-                                            End : InProgress;
-                                }
-                                if (itemWritten instanceof HttpHeaders) {
-                                    return End;
-                                }
-                                return InProgress;
+            flushStrategy = new SplittingFlushStrategy(connection.defaultFlushStrategy(),
+                    new FlushBoundaryProvider() {
+                        private long contentLength;
+                        // h2 may return a single HttpResponseMetaData for an empty response in some scenarios,
+                        // otherwise a trailers object will be included to indicate the end because content-length isn't
+                        // mutually exclusive from trailers in h2.
+                        private final boolean trailersUnlessEmpty = protocol().major() > 1;
+                        @Override
+                        public FlushBoundary detectBoundary(@Nullable final Object itemWritten) {
+                            if (itemWritten instanceof HttpResponseMetaData) {
+                                final HttpResponseMetaData metadata = (HttpResponseMetaData) itemWritten;
+                                contentLength = getContentLength(metadata);
+                                // The content length maybe unknown at this point (e.g. 204 response) but then later
+                                // determined to be 0. In that case we should conservatively use End and rely upon
+                                // adjustForMissingBoundaries to accommodate if more data comes. Otherwise the
+                                // FlushStrategy may not trigger a flush (e.g. flushOnEnd) and if so the response
+                                // won't actually be written.
+                                return contentLength > 0 || (contentLength < 0 &&
+                                        (!emptyMessageBody(metadata) &&
```

Comment:
We tend to add `content-length: 0` for the empty body optimization in h2, but figured I would add `emptyMessageBody` to be safe.